service/history/workflow/retry.go

```diff
@@ -32,26 +32,63 @@ import (
 	failurepb "go.temporal.io/api/failure/v1"
 
 	"go.temporal.io/server/common/backoff"
-	"go.temporal.io/server/common/primitives/timestamp"
 )
 
 func getBackoffInterval(
 	now time.Time,
-	expirationTime time.Time,
-	currentAttemptCounterValue int32,
+	currentAttempt int32,
 	maxAttempts int32,
 	initInterval *time.Duration,
 	maxInterval *time.Duration,
+	expiration *time.Time,
 	backoffCoefficient float64,
 	failure *failurepb.Failure,
 	nonRetryableTypes []string,
 ) (time.Duration, enumspb.RetryState) {
 
-	// Sanity check to make sure currentAttemptCounterValue started with 1.
-	if currentAttemptCounterValue < 1 {
-		currentAttemptCounterValue = 1
+	var initIntervalDuration *time.Duration
+	if initInterval != nil && *initInterval > 0 {
+		initIntervalDuration = initInterval
 	}
 
+	var maxIntervalDuration *time.Duration
+	if maxInterval != nil && *maxInterval > 0 {
+		maxIntervalDuration = maxInterval
+	}
+
+	var expirationTime *time.Time
+	if expiration != nil && !(*expiration).IsZero() {
+		expirationTime = expiration
+	}
+
+	return getBackoffIntervalWrapper(
+		now,
+		currentAttempt,
+		maxAttempts,
+		initIntervalDuration,
+		maxIntervalDuration,
+		expirationTime,
+		backoffCoefficient,
+		failure,
+		nonRetryableTypes,
+	)
+}
+
+// TODO treat 0 as 0, not infinite
+// TODO use getBackoffIntervalWrapper for retry calculation, deprecate the 0 -> nil conversion
+
+func getBackoffIntervalWrapper(
```

Comment:
This is not a wrapper. This is wrappable. I think `getBackoffIntervalInternal` or `getBackoffIntervalNilable` are better names here.